### DISCUSSION TITLE
string/intern

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#119] [Announcements]
# string/intern

Java程序员进阶之路，小白的零基础Java教程，从入门到进阶，深入解析String.intern()方法

https://tobebetterjavaer.com/string/intern.html

**Comment by dark-eternity** [comment]
String s1 = new String("二哥") + new String("三妹");
"二哥"、"三妹" 存在了常量池跟堆，然后这个+又是怎么进行的，发生了什么，为何"二哥三妹"只是存在了堆里，没进常量池

**Reply by lcz857** [reply]
这里底层是new了一个StringBuilder，然后通过append方法接上去的，append方法实际上就是对value字符数组进行操作，然后将结果调用toString方法（new String（char value[], int offset, int count）），里面就是Arrays.copyOfRange方法，直接写入了String类里value属性的值，这时不会在常量池创建字面量的对象的。我理解为new String构造器里面传字面量就会在常量池创建相应对象，就相当于把字面量里的值先声明定义了一下，此时调用intern会直接返回常量池里字面量的引用；而new String构造器传的不是字面量的话常量池就没有对应的值，此时调用intern会创建一个引用指向堆中String对象，此时对象的引用和intern返回的值是一样的。

**Reply by Tim1Wu** [reply]
其实简单一句话 StringBuilder创建的是可变长的字符串变量 可变长就意味着不可能是常量，所以自然不会在常量池创建

**Comment by TH-zhou** [comment]
我有一个问题，看美团那篇技术文章，说的是new String的时候不会放在字符串常量池中，只有调intern方法不存在字符串常量池中就会放入；但是看你这篇的时候说new的时候就直接放入常量池了，所以，我现在还是不太清楚，到底啥时候，啥动作才会放入字符串常量池中

**Reply by javabcde** [reply]
那篇文章你没仔细看吧 你去搜一下这段话
String s = new String("1"); 第一句代码，生成了2个对象。常量池中的“1” 和 JAVA Heap 中的字符串对象。

**Comment by MaggieLee01** [comment]
好奇一个问题，文章说“ Java 7 之后呢，由于字符串常量池放在了堆中，执行 String.intern() 方法的时候，如果对象在堆中已经创建了，字符串常量池中就不需要再创建新的对象了，而是直接保存堆中对象的引用，也就节省了一部分的内存空间。”，JAVA8中，字符串常量池不在堆中了，执行 String.intern() 方法的时候字符串常量会重新创建一个对象吗

**Reply by tough-road** [reply]
JDK8常量池不还在堆中嘛

**Comment by ThePHOENlX** [comment]
请问为什么在这条语句后String s1 = new String("二哥") + new String("三妹");，字符串常量池保存的是堆上对象的引用呢？

**Reply by JPL1988** [reply]
Returns a canonical representation for the string object.
A pool of strings, initially empty, is maintained privately by the class String.
When the intern method is invoked, if the pool already contains a string equal to this String object as determined by the equals(Object) method, then the string from the pool is returned. Otherwise, this String object is added to the pool and a reference to this String object is returned.
It follows that for any two strings s and t, s. intern() == t. intern() is true if and only if s. equals(t) is true.
All literal strings and string-valued constant expressions are interned. String literals are defined in section @jls 3.10.5 of the The Java Language Specification.
返回值:
a string that has the same contents as this string, but is guaranteed to be from a pool of unique strings.

根据java的api注释还有美团的文章https://tech.meituan.com/2014/03/06/in-depth-understanding-string-intern.html，我觉得这一句返回的就是在常量池的字符串引用，而不是常量池在指向堆内存的引用，图画错了可能

**Reply by LiSeries** [reply]
String.intern() 方法的目的是确保字符串常量池中的唯一性。当调用 intern() 方法时，它会检查常量池中是否已经存在该字符串：
如果常量池中已经存在该字符串，则返回池中该字符串的引用。
如果常量池中没有该字符串，则将其添加到常量池中，并返回这个新添加字符串的引用。

**Reply by LiSeries** [reply]
回复咋撤回不了，当我没说过sorry。

**Reply by chulinbai** [reply]
JDK6：当调用intern方法的时候，如果字符串常量池先前已创建出该字符串对象，则返回常量池中的该字符串的引用。否则，将此字符串对象添加到字符串常量池中，并且返回该字符串对象的引用。

JDK6+：当调用intern方法的时候，如果字符串常量池先前已创建出该字符串对象，则返回常量池中的该字符串的引用。否则，如果该字符串对象已存在与Java堆中，则将堆中对此对象的引用添加到字符串常量池中，并且返回该引用；如果堆中不存在，则在常量池中创建该字符串并返回其引用。

确实是引用。
